Add docker build capability.  Add "latest" as a choice for update-janet.  Improve README.md

MikeBeller opened the pull request.

MikeBeller committed:
Trying out docker for more reliable builds

MikeBeller committed:
add build-docker, update readme

MikeBeller committed:
allow 'latest' as a version in update-janet

veqqq commented:
I had to make a few changes (reverted some path names to use `".."` as the first element and `import` to use `./playground`).  N.b. zip via: https://api.github.com/repos/MikeBeller/janet-playground/zipball/pull/13/head

I don't understand why you changed them, so I'm not sure if this is the best approach. I also don't understand why `install` was fine without `os/rename` needing them.

-----

I didn't touch emsdk or anything (since already installed).

I did `./build update-janet latest` but the import was broken, so I changed `playground`'s `(import "playground" :as play)` to `(import "./playground" :as play)`. 

```
./playground update-janet latest
GETTING: https:/github.com/janet-lang/janet/releases/latest/download/janet.c
GETTING: https:/github.com/janet-lang/janet/releases/latest/download/janet.h
veqqq:~/Desktop/MikeBeller-janet-playground-68b02ed$ ./playground build
emcc: error: janet/janet.c: No such file or directory ("janet/janet.c" was expected to be an input file, based on the commandline arguments provided)
Error running emcc.  Build failed.
```

I confirmed, it is the correct janet.c (1.41). I added ".." to build's definition like the old code:

```
            (path/join ".." "janet" "janet.c")
            (path/join ".." "janet" "play.c")
```

But:

```
Desktop/MikeBeller-janet-playground-68b02ed$ ./playground build
emcc: error: specified output file (build/janet.js) is in a directory that does not exist
Error running emcc.  Build failed.
```
so I also changed:

```
  (def result
    (try
      (do
        (os/execute
          (array/join emcc @["-O2" "-o" "janet.js"
            (path/join ".." "janet" "janet.c")
            (path/join ".." "janet" "play.c")
            (string "-I" "janet")
            "--embed-file" (path/join ".." "build" "fmt.jimage")
            "-s" "EXPORTED_FUNCTIONS=['_run_janet']"
            "-s" "ALLOW_MEMORY_GROWTH=1"
            "-s" "AGGRESSIVE_VARIABLE_ELIMINATION=1"
            "-s" "EXPORTED_RUNTIME_METHODS=['ccall','cwrap']"
            "-o" (path/join ".." "build" "janet.js")]
          )
          :p))
```

then install, then serve and everything works (verified the version via serve.

sogaiu commented:
Below is a diff / patch that is mostly what @veqqq mentioned (plus a small change to a `defn` [1]).

With the patch applied I had success fetching the latest (via `update-latest`) and building with a local emsdk 4.0.5 (via `build`).

(Sorry I don't have a docker setup to test with docker :sweat_smile:)

```diff
diff --git a/playground b/playground
index fbd8359..7262fdf 100755
--- a/playground
+++ b/playground
@@ -1,6 +1,6 @@
 #!/usr/bin/env janet
 
-(import "playground" :as play)
+(import ./playground :as play)
 
 (def args (dyn :args))
 
diff --git a/playground.janet b/playground.janet
index cb8decb..81616d0 100644
--- a/playground.janet
+++ b/playground.janet
@@ -21,11 +21,11 @@
   (def env (require "spork/fmt"))
   (spit "fmt.jimage" (make-image env)))
 
-(defn build (use-docker)
+(defn build
   ```Build wasm module.  Emsdk must be installed and runnable in the current
   shell (on linux, for example, `source emsdk_env.sh` in the emsdk directory,)
   else docker must be installed and running, and the user must have docker privileges.```
-  []
+  [use-docker]
   (def stat (os/stat "build"))
   (when (nil? stat)
     (os/mkdir "build"))
@@ -44,15 +44,15 @@
       (do
         (os/execute
           (array/join emcc @["-O2" "-o" "janet.js"
-            (path/join "janet" "janet.c")
-            (path/join "janet" "play.c")
+            (path/join ".." "janet" "janet.c")
+            (path/join ".." "janet" "play.c")
             (string "-I" "janet")
-            "--embed-file" (path/join "build" "fmt.jimage")
+            "--embed-file" (path/join ".." "build" "fmt.jimage")
             "-s" "EXPORTED_FUNCTIONS=['_run_janet']"
             "-s" "ALLOW_MEMORY_GROWTH=1"
             "-s" "AGGRESSIVE_VARIABLE_ELIMINATION=1"
             "-s" "EXPORTED_RUNTIME_METHODS=['ccall','cwrap']"
-            "-o" (path/join "build" "janet.js")]
+            "-o" (path/join ".." "build" "janet.js")]
           )
           :p))
       ([err] (eprint "Can't run emcc.  Ensure emcc is installed and in your path, and emsdk_env.sh has been sourced into your current environment"))))
```

---

[1] Without this change I think the string intended as a docstring doesn't end up as one.

MikeBeller commented:
That’s strange. If you’ll note in the new instructions you (1) clone the
repo, (2) cd into it (3) jpm build, (4) jpm install, (5) ./playground
update-janet latest (6) ./playground build

You definitely shouldn’t have to change those paths. It’s supposed to work
from the main directory not from the build directory so ../ is just wrong
now.

Let me look at it again as I must have failed to change it for the non
docker case.

Mike


On Fri, Nov 28, 2025 at 9:36 PM sogaiu ***@***.***> wrote:

> *sogaiu* left a comment (MikeBeller/janet-playground#13)
> <https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3590917692>
>
> Below is a diff / patch that is mostly what @veqqq
> <https://github.com/veqqq> mentioned (plus a small change to a defn).
>
> With the patch applied I had success fetching the latest (via
> update-latest) and building with a local emsdk 4.0.5 (via build).
>
> (Sorry I don't have a docker setup to test with docker 😅)
>
> diff --git a/playground b/playground
> index fbd8359..7262fdf 100755--- a/playground+++ b/playground@@ -1,6 +1,6 @@
>  #!/usr/bin/env janet
>  -(import "playground" :as play)+(import ./playground :as play)
>
>  (def args (dyn :args))
>  diff --git a/playground.janet b/playground.janet
> index cb8decb..81616d0 100644--- a/playground.janet+++ b/playground.janet@@ -21,11 +21,11 @@
>    (def env (require "spork/fmt"))
>    (spit "fmt.jimage" (make-image env)))
>  -(defn build (use-docker)+(defn build
>    ```Build wasm module.  Emsdk must be installed and runnable in the current
>    shell (on linux, for example, `source emsdk_env.sh` in the emsdk directory,)
>    else docker must be installed and running, and the user must have docker privileges.```-  []+  [use-docker]
>    (def stat (os/stat "build"))
>    (when (nil? stat)
>      (os/mkdir "build"))@@ -44,15 +44,15 @@
>        (do
>          (os/execute
>            (array/join emcc @["-O2" "-o" "janet.js"-            (path/join "janet" "janet.c")-            (path/join "janet" "play.c")+            (path/join ".." "janet" "janet.c")+            (path/join ".." "janet" "play.c")
>              (string "-I" "janet")-            "--embed-file" (path/join "build" "fmt.jimage")+            "--embed-file" (path/join ".." "build" "fmt.jimage")
>              "-s" "EXPORTED_FUNCTIONS=['_run_janet']"
>              "-s" "ALLOW_MEMORY_GROWTH=1"
>              "-s" "AGGRESSIVE_VARIABLE_ELIMINATION=1"
>              "-s" "EXPORTED_RUNTIME_METHODS=['ccall','cwrap']"-            "-o" (path/join "build" "janet.js")]+            "-o" (path/join ".." "build" "janet.js")]
>            )
>            :p))
>        ([err] (eprint "Can't run emcc.  Ensure emcc is installed and in your path, and emsdk_env.sh has been sourced into your current environment"))))
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3590917692>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM7HKQRGGXIV74RDIARAPT37EBDRAVCNFSM6AAAAACNQI2ILOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJQHEYTONRZGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sogaiu commented:
I think at least in my case I didn't follow the new instructions precisely:

```
(1) clone the repo,
(2) cd into it
(3) jpm build,
(4) jpm install,
(5) ./playground update-janet latest
(6) ./playground build
```

Will investigate and get back to you.

sogaiu commented:
I tried [the steps above](https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3590938370) but they did not work for me because `spork` was not found.  IIUC, some kind of `jpm deps` may be necessary.  

Below are more details about a subsequent attempt (with no patching).

```
cd ~/src
source ~/src/emsdk/emsdk_env.sh
git clone https://github.com/janet-playground
cd janet-playground
git checkout docker
jpm --local deps
jpm --local install
jpm --local janet playground update-janet latest
# step below fails
jpm --local janet playground build
emcc: error: janet/janet.c: No such file or directory ("janet/janet.c" was expected to be an input file, based on the commandline arguments provided)
Error running emcc.  Build failed.
```

sogaiu commented:
I wonder what the benefit(s) is(are) in changing the import path in the `playground` executable Janet script from `./playground` to `playground` (or `"plaground"`).

FWIW, so far in typical use, I use the `playground` Janet script directly from the repository root [1].  If the path in the import form is kept as `./playground`:

```janet
(import ./plaground) # or (import "./playground")
```

then IIUC users don't need to run the `jpm install` step and can rely on the version of `playground.janet` in the repository root being the one that is used.

In constrast, if a switch is made to `(import playground :as play)` (or `(import "playground" :as play)`), we are now faced with having to install as well as potentially facing an expectation mismatch upon updates to `playground.janet` (if we fail to remember to `jpm install` again).

Not sure I articulated that very well...please let me know :sweat_smile: 

Any thoughts?

---

[1] IIUC, since the `janet` subdirectory in the repository contains files that `playground` needs when running certain commands, I think we need to run the `playground` Janet script from the repository root anyway (and perhaps for other reasons)...though may be I'm off on that.

MikeBeller commented:
Yes I am wondering the same. I prefer not needing to install
Janet-playground ever. But there are some dependencies like osprey, sh,
etc. how should we make sure those are accessible without jpm install?

On Sat, Nov 29, 2025 at 12:41 AM sogaiu ***@***.***> wrote:

> *sogaiu* left a comment (MikeBeller/janet-playground#13)
> <https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3591017471>
>
> I wonder what the benefit(s) is(are) in changing the import path from
> ./playground to playground (or "plaground") might be in the playground
> executable Janet script.
>
> FWIW, so far in typical use, I use the playground Janet script directly
> from the repository root. If the path in the import form is kept as
> ./playground:
>
> (import ./plaground) # or (import "./playground")
>
> then IIUC users don't need to run the jpm install step and can rely on
> the version of playground.janet in the repository root being the one that
> is used.
>
> In constrast, if a switch is made to (import playground :as play) (or (import
> "playground" :as play)), we are now faced with having to install as well
> as potentially facing an expectation mismatch upon updates to
> playground.janet (if we fail to remember to jpm install again).
>
> Not sure I articulated that very well...please let me know 😅
>
> Any thoughts?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3591017471>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM7HKT5NVIJJSR7VBSKRUD37EWYXAVCNFSM6AAAAACNQI2ILOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJRGAYTONBXGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sogaiu commented:
I believe `jpm deps` is sufficient.

I tend to use `jpm --local deps` though to keep things isolated to the specific project.  IIUC, a `jpm_tree` subdirectory ends up getting populated appropriately.

FWIW, in the [steps mentioned above](https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3590943262) `--local` was used in a few places:

```
jpm --local deps
jpm --local install      # I think we can avoid this
jpm --local janet playground update-janet latest
# step below fails
jpm --local janet playground build
```

MikeBeller committed:
Fix local build, update readme, jpm local

MikeBeller commented:
> I believe `jpm deps` is sufficient.
> 
> I tend to use `jpm --local deps` though to keep things isolated to the specific project. IIUC, a `jpm_tree` subdirectory ends up getting populated appropriately.
> 
> FWIW, in the [steps mentioned above](https://github.com/MikeBeller/janet-playground/pull/13#issuecomment-3590943262) `--local` was used in a few places:
> 
> ```
> jpm --local deps
> jpm --local install      # I think we can avoid this
> jpm --local janet playground update-janet latest
> # step below fails
> jpm --local janet playground build
> ```

Ok @sogaiu -- I did all the things. 

- Fixed local emcc build (at least for me)
- switched to "./playground"
- updated readme to reflect the new instructions (jpm --local and no jpm install).

Please test and let me know if everything works for you.  (I tested docker so you don't need to do that if you don't have docker).  If all is well I will merge.

sogaiu commented:
@MikeBeller Thanks!

I pulled the latest changes, took a look through the README and gave the instructions a try for a local emsdk installation (I also tried the `serve` command and tested some of the served pages via a web browser).

It went pretty well :+1: 

---

I think there are a couple of things may be worth tweaking.  One seems somewhat important and the other one is relatively minor.

1. IIUC, the README no longer mentions the [`install` command for `playground` itself](https://github.com/MikeBeller/janet-playground/pull/13/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L23) (not `jpm`).

    When @veqqq and I were investigating before with different versions of `janet.c`, I failed to run the `playground install` command which led to some confusion as older versions of `janet.wasm` and `janet.js` were being served instead of freshly built versions.  When I just tested the latest changes the `install` command was still needed (which is fine I think).

    I think it would be less confusing to continue to point out `playground`'s `install` command in the README.

2. On a minor note, I think there is a slight problem with the way the `build` function was changed having to do with the order of the argument tuple and the docstring.  With the way it is in [here](https://github.com/MikeBeller/janet-playground/pull/13/files#diff-57649a39cc4f4559a36cefbf0a30be94344aee2900d5e692efa3bbf53ad79794L23-L26), the string that looks to be intended as a docstring will not be treated by `defn` as a docstring.  I think the following would work better:

```janet
(defn build
  ``Build wasm module.  Emsdk must be installed and runnable in the current
  shell (on linux, for example, `source emsdk_env.sh` in the emsdk directory,)
  else docker must be installed and running, and the user must have docker privileges.``
  [use-docker]
  (def stat (os/stat "build"))
```

---

May be @veqqq will get a chance to take a look at the most recent changes too.

MikeBeller committed:
fix doc comment, readme.

MikeBeller commented:
> @MikeBeller Thanks!
> 
> I pulled the latest changes, took a look through the README and gave the instructions a try for a local emsdk installation (I also tried the `serve` command and tested some of the served pages via a web browser).
> 
> It went pretty well 👍
> 
> I think there are a couple of things may be worth tweaking. One seems somewhat important and the other one is relatively minor.
> 
> 1. IIUC, the README no longer mentions the [`install` command for `playground` itself](https://github.com/MikeBeller/janet-playground/pull/13/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L23) (not `jpm`).
>    When @veqqq and I were investigating before with different versions of `janet.c`, I failed to run the `playground install` command which led to some confusion as older versions of `janet.wasm` and `janet.js` were being served instead of freshly built versions.  When I just tested the latest changes the `install` command was still needed (which is fine I think).
>    I think it would be less confusing to continue to point out `playground`'s `install` command in the README.
> 2. On a minor note, I think there is a slight problem with the way the `build` function was changed having to do with the order of the argument tuple and the docstring.  With the way it is in [here](https://github.com/MikeBeller/janet-playground/pull/13/files#diff-57649a39cc4f4559a36cefbf0a30be94344aee2900d5e692efa3bbf53ad79794L23-L26), the string that looks to be intended as a docstring will not be treated by `defn` as a docstring.  I think the following would work better:
> 
> ```janet
> (defn build
>   ``Build wasm module.  Emsdk must be installed and runnable in the current
>   shell (on linux, for example, `source emsdk_env.sh` in the emsdk directory,)
>   else docker must be installed and running, and the user must have docker privileges.``
>   [use-docker]
>   (def stat (os/stat "build"))
> ```
> 
> May be @veqqq will get a chance to take a look at the most recent changes too.

Thanks -- good catch @sogaiu  -- I fixed those two items.  Let me know if you see anything else.  @veqqq let us know if you have comments. -M

veqqq commented:
The current instructions work for me!

MikeBeller commented:
> The current instructions work for me!

Great.  Merging.  Thanks @sogaiu and @veqqq

sogaiu commented:
Thanks @MikeBeller and @veqqq :)
